### ISSUE TITLE
Using `flash_attention_2` raises a `ValueError` for `padding_side`

### ISSUE THREAD

**Issue by ylkhayat** [hongshi97/CAD#2]
Thank you for developing this!

## Context
Due to lenghty computation time and in order to speed things up, I thought about using the `flash_attention_2` and smaller floating points `torch.float16`  as mentioned here https://huggingface.co/docs/transformers/main_classes/pipelines#pipeline-fp16-inference.

> I am also using `model_name = mistralai/Mistral_Instruct_v0.3`

```python
class CAD:
    def __init__(self, model_name: str, device: Union[int,str] = 0):
        self.model = AutoModelForCausalLM.from_pretrained(model_name, device_map=device_map, use_cache=True, attn_implementation="flash_attention_2", torch_dtype=torch.float16)
        self.tokenizer = AutoTokenizer.from_pretrained(model_name)
```

## Issue
But I kept getting a specific error about using `padding_right` in transformers, you can find it here 

https://github.com/huggingface/transformers/blob/4e90b99ed916300b80bac9db793f2a96b2a87122/src/transformers/models/mistral/modeling_mistral.py#L841-L849

```bash 
ValueError: You are attempting to perform batched generation with padding_side='right' this may lead to unexpected behaviour for Flash Attention version of Mistral. Make sure to  call `tokenizer.padding_side  = 'left'` before tokenizing the input.
```

I am wondering if this can be solved by simply always concatenating the accept token to the end and not worry about the padding and using `padding_side="left"` to the tokenizer, ie `self.tokenizer = AutoTokenizer.from_pretrained(model_name, padding_side="left")`.

### Details

This happens exactly when one input from the batch finishes before the rest. Then, the error propagates from the consequences of these lines, because at some point, our `attention_mask`s will not contain 1s at the end of each sequence. Which triggers the transformer line to break

```py
input_ids = torch.cat([input_ids, tokens_to_add.unsqueeze(-1)], dim=-1)
attention_mask = torch.cat([attention_mask, unfinished_sents.unsqueeze(-1)], dim=-1)
input_ids_with_contexts = torch.cat([input_ids_with_contexts, tokens_to_add.unsqueeze(-1)], dim=-1)
attention_mask_with_contexts = torch.cat([attention_mask_with_contexts, unfinished_sents.unsqueeze(-1)], dim=-1)
```

due to this condition from the transformers [file](https://github.com/huggingface/transformers/blob/4e90b99ed916300b80bac9db793f2a96b2a87122/src/transformers/models/mistral/modeling_mistral.py#L843C17-L843C96)

```py
is_padding_right = attention_mask[:, -1].sum().item() != input_tensor.size()[0]
```

**Comment by hongshi97** [comment]
Hi,

Thank you for sharing this issue. While I'm not completely certain, it seems likely that the error is related to the `padding_side` setting in the tokenizer when using flash_attention_2 with Mistral.

FlashAttention_2 often requires padding_side to be set to 'left' for proper alignment during batch processing. I suggest trying the following adjustment to your tokenizer:
`tokenizer.padding_side = 'left'`

This configuration is recommended in cases similar to yours and is discussed in the Hugging Face Transformers repository ([link to discussion](https://github.com/huggingface/transformers/issues/26877)).

Additionally, since this work builds on the original Context-Aware Decoding paper, it might also be helpful to raise this question in the original implementation’s repository. You can find the repository here: [Context-Aware Decoding Original Repository](https://github.com/xhan77/context-aware-decoding).

If this doesn’t resolve the issue, it might help to verify other aspects of the environment, such as library versions.

Looking forward to hearing back from you to ensure this gets resolved.

**Comment by ylkhayat** [comment]
Yes, I did manage to solve this issue through setting the `padding_side = 'left'` which encouraged along the way some other changes. I will gladly share the changes below with a comment `— Issue Relevant —` to easily identify lines which concern the issue and provide clear steps to solve the issue.

## Changes

### Initialization
I currently initialize the model as follows in order to use the `flash_attention_2` and `torch.float16`.

```py
class CAD:
    def __init__(self, model_name: str, device: Union[int,str] = 0):
        device_map = torch.device(f"cuda:{device}" if torch.cuda.is_available() else "cpu")
        # — Issue Related —
        self.model = AutoModelForCausalLM.from_pretrained(model_name, device_map=device_map, use_cache=True, attn_implementation="flash_attention_2", torch_dtype=torch.float16)
        # — Issue Related —
        self.tokenizer = AutoTokenizer.from_pretrained(model_name, padding_side="left")
        self.device = device_map
        # — Issue Related —
        self.tokenizer.pad_token = self.tokenizer.eos_token
```

### Generation
Then, I define the `model_kwargs` along with the preparation of the generation

#### Without Context
```py
# Before the Generation Loop
model_kwargs = {
    "use_cache": True,
    "attention_mask": attention_mask
}

# In the Generation Loop
model_inputs = self.model.prepare_inputs_for_generation(input_ids, **model_kwargs)

outputs = self.model(**model_inputs, return_dict=True)

# — Issue Related —
model_kwargs["attention_mask"] = torch.cat([model_kwargs["attention_mask"], torch.ones((batch_size, 1), device=self.device)], dim=-1)
model_kwargs["past_key_values"] = outputs.past_key_values
```

#### With Context
```py
# Before the Generation Loop
model_kwargs_with_contexts = {
    "use_cache": True,
    "attention_mask": attention_mask_with_contexts
}

# In the Generation Loop
model_inputs_with_contexts = self.model.prepare_inputs_for_generation(input_ids_with_contexts, **model_kwargs_with_contexts)

outputs_with_contexts = self.model(**model_inputs_with_contexts, return_dict=True)

 # — Issue Related —
model_kwargs_with_contexts["attention_mask"] = torch.cat([model_kwargs_with_contexts["attention_mask"], torch.ones((batch_size, 1), device=self.device)], dim=-1)
model_kwargs_with_contexts["past_key_values"] = outputs_with_contexts.past_key_values
```

### Sync

Then, I only do the following to update my input ids.

```py
# — Issue Related —
input_ids = torch.cat([input_ids, next_token.unsqueeze(-1)], dim=-1)
# — Issue Related —
input_ids_with_contexts = torch.cat([input_ids_with_contexts, next_token.unsqueeze(-1)], dim=-1)
```

Finally, this is how the loop wraps.

```py
# — Issue Related —
cur_len += 1
for i, token in enumerate(next_token.tolist()):
    if unfinished_sents[i] == 1:
        generated_tokens[i].append(token)
            # — Issue Related —
            if unfinished_sents[i] == 1 and token == self.tokenizer.eos_token_id:
                unfinished_sents[i] = 0
                sent_lengths[i] = cur_len
            # — Issue Related —
            if unfinished_sents.max() == 0:
                break

```

## Take-away
I would gladly provide a pull request if the changes mentioned above is needed and requested. Yet, I experimented this code multiple times and it seems to be working nicely and quickly.

Once again, here's my takes on the changes:

### Issue Relevant

- Main drive of these changes is the adaptability to the `left` padding nature of decoders - which arose from the previous mentioned error while trying to use `flash_attention_2`. And it was motivated by the fact that we usually pad on the left for batches and handle the generation from there.
- This implementation tackles the issue of handling paddings manually since the tokenizer takes care of them only once in the beginning, and we do not have to worry about them ever again.
- This code was only tested on mistral models and might bring potential issues for the rest.

### Issue Irrelevant
- These changes bring lightning fast generation through `caching` and using the `past_key_values`.

Finally, I believe that this code might be buggy but please if any breaking points are visible in sight guide me to them. Thanks again for the effort to bring this piece of code together, my suggestions are only humble suggestions to improve the overall experience of running `CAD` and testing it.
